scipy/stats/_morestats.py

```diff
@@ -1309,29 +1321,40 @@ def _all(x):
         message = ("The `optimizer` argument of `boxcox_normmax` must return "
                    "an object containing the optimal `lmbda` in attribute `x`.")
         raise ValueError(message)
-    else:
-        # Test if the optimal lambda causes overflow
+    elif ymax is None or not np.isinf(ymax):  # adjust the final lambda
         x = np.asarray(x)
-        max_x = np.max(x, axis=0)
-        istransinf = np.isinf(special.boxcox(max_x, res))
-        dtype = x.dtype if np.issubdtype(x.dtype, np.floating) else np.float64
-        if np.any(istransinf):
-            warnings.warn(
-                f"The optimal lambda is {res}, but the returned lambda is "
-                f"the constrained optimum to ensure that the maximum of the "
-                f"transformed data does not cause overflow in {dtype}.",
-                stacklevel=2
-            )
-
-            # Return the constrained lambda to ensure the transformation
-            # does not cause overflow
-            ymax = np.finfo(dtype).max / 100  # 100 is the safety factor
-            constrained_res = _boxcox_inv_lmbda(max_x, ymax)
-
-            if isinstance(res, np.ndarray):
-                res[istransinf] = constrained_res
-            else:
-                res = constrained_res
+        xmax = np.max(x, axis=0)
+        ymax_res = special.boxcox(xmax, res)
+
+        if ymax is None:
+            # Test if the optimal lambda causes overflow
+            istransinf = np.isinf(ymax_res)
```

Comment:
```suggestion
            is_overflow = np.isinf(ymax_res)
```
If `xmax=np.inf`, then it was not caused by overflow.

Reply:
I try to modify the test introduced in #18764 as the following, but for `method='mle'` it raises `BracketError` not `ValueError`. So I think maybe we need to check the input at first (raise `ValueError`) and also remove the `try-except` part.

```python
@pytest.mark.parametrize(
    "bad_x", [
        np.array([1, -42, 12345.6]),
        np.array([np.nan, 42, 1]),
        np.array([np.inf, 42, 1]), # add np.inf
    ]
)
@pytest.mark.parametrize(
    "method", ["pearsonr", "mle"] # add "mle" since default is "pearsonr"
)
def test_negative_x_value_raises_error(bad_x, method):
    """Test boxcox_normmax raises ValueError if x contains non-positive values."""
    message = "only positive, finite, real numbers"
    with pytest.raises(ValueError, match=message):
        stats.boxcox_normmax(bad_x, method=method)
# scipy.optimize._optimize.BracketError: The algorithm terminated without finding a valid bracket. Consider trying different initial points.
```

cc @mdhaber @fkiraly